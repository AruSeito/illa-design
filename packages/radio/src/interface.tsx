import { HTMLAttributes, ChangeEvent } from "react"

export type RadioColorScheme =
  string
  | "white"
  | "blackAlpha"
  | "gray"
  | "grayBlue"
  | "red"
  | "orange"
  | "yellow"
  | "green"
  | "blue"
  | "cyan"
  | "purple"

export interface RadioContextProps {
  name?: string
  value?: any
  disabled?: boolean
  checked?: boolean
  defaultChecked?: boolean

  colorScheme?: RadioColorScheme
  onChange?: (value: any, event: ChangeEvent) => void;
}

export type RadioOwnProps = HTMLAttributes<HTMLDivElement> & RadioContextProps

export interface RadioProps extends RadioOwnProps {}

export type RadioOwnGroupProps = HTMLAttributes<HTMLDivElement> & RadioGroupContextProps

export interface RadioGroupProps extends RadioOwnGroupProps {}

interface optionsArr {
  label: string
  value: string
  disabled?: boolean
}[]

export interface RadioGroupContextProps {
  name?: string
  value?: any
  colorScheme?: RadioColorScheme
  disabled?: boolean
  defaultValue?: any
  options?: string[] | optionsArr[]
  direction?: "vertical" | "horizontal"
  spacing?: string | number


  onChange?: (value: any, event: ChangeEvent) => void;
  onChangeValue?: (value: any, event: ChangeEvent) => void;
}