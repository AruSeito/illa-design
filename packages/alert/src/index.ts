export * from "./alert"
export * from "./interface"
