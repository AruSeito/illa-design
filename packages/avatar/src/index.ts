export * from "./avatar"
export * from "./interface"
export * from "./avatar-group"