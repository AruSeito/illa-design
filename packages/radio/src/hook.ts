export function useMergeValue<T>(
  defaultStateValue: T,
  props?: {
    defaultValue?: T;
    value?: T;
  }
) {
  const { defaultValue, value } = props || {};

  return value !== undefined ? value : defaultValue !== undefined ? defaultValue : defaultStateValue;
}

