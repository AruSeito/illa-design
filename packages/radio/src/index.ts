export * from "./radio"
export * from "./interface"
export * from "./radio-group"