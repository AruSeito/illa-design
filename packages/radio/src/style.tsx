import { globalColor, illaPrefix } from "@illa-design/theme"
import { RadioColorScheme, RadioProps } from "./interface"
import { SerializedStyles } from "@emotion/serialize"
import { css } from "@emotion/react"

// default radio
export function applyRadioSize(colorScheme: RadioColorScheme) {
  return css`
    -webkit-appearance: none;
    -moz-appearance: none;
    appearance: none;
    border-radius: 50%;
    margin: auto 6px auto auto;
    width: 16px;
    height: 16px;
    border: solid 2px ${globalColor(`--${illaPrefix}-gray-08`)};
    cursor: pointer;
    transition: 0.15s all linear;
    &:hover {
      border-color: ${globalColor(`--${illaPrefix}-${colorScheme}-06`)};
    }
    &:disabled {
      cursor: not-allowed;
      border-color: ${globalColor(`--${illaPrefix}-gray-08`)};
      background-color:  ${globalColor(`--${illaPrefix}-gray-09`)};
    }
    
    &:checked {
      border: 4px solid ${globalColor(`--${illaPrefix}-${colorScheme}-01`)};
      &:hover {
        border-color: ${globalColor(`--${illaPrefix}-${colorScheme}-02`)};
      }
      &:disabled {
        border-color: ${globalColor(`--${illaPrefix}-${colorScheme}-05`)};
      }
    }
  `
}

export function applyRadioContainer() {
  return css`
    display: inline-flex;
    overflow: hidden;
    word-wrap: break-word;
    white-space: nowrap;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    //padding: 4px;
    cursor: pointer;
    color: ${globalColor(`--${illaPrefix}-gray-02`)};
    &:disabled {
      cursor: not-allowed;
    }
  `
}

export function applyMergeCss(props: RadioProps): SerializedStyles {
  const currentDisabled = props.disabled ?? false

  return css`
    display: inline-flex;
    overflow: hidden;
    word-wrap: break-word;
    white-space: nowrap;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    color: ${globalColor(`--${illaPrefix}-gray-02`)};
    cursor: ${currentDisabled ? 'not-allowed': 'pointer'};
  `
}

export function applyRadioContainerHorizontal(spacing: string|number): SerializedStyles {
  const currentSpacing = typeof spacing === "string" ? spacing : `${spacing}px`

  return css`
    display: inline-flex;
    vertical-align: middle;
    flex-direction: row;
    align-items: center;
    gap: ${currentSpacing};
  `
}

export function applyRadioContainerVertical(spacing: string|number): SerializedStyles {
  const currentSpacing = typeof spacing === "string" ? spacing : `${spacing}px`

  return css`
    display: inline-flex;
    vertical-align: middle;
    flex-direction: column;
    align-items: flex-start;
    gap: ${currentSpacing};
    margin-bottom: ${currentSpacing};
  `
}