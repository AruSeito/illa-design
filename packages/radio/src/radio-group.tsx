/** @jsxImportSource @emotion/react */
import { ChangeEvent, createContext, forwardRef } from "react"
import { RadioGroupContextProps, RadioGroupProps } from "./interface"
import { SerializedStyles } from "@emotion/react"
import { Radio } from "./radio"
import {
  applyRadioContainerHorizontal,
  applyRadioContainerVertical,
} from "./style"
import { useMergeValue } from "./hook"

function isArray(obj: any) {
  return Object.prototype.toString.call(obj) === "[object Array]"
}

interface extraProps {
  onChangeValue?: (value: any, event: ChangeEvent) => void
}

export const RadioGroupContext = createContext<
  (RadioGroupContextProps<any> & extraProps) | undefined
>(undefined)

export const RadioGroup = forwardRef<HTMLDivElement, RadioGroupProps<any>>(
  (props, ref) => {
    const {
      children,
      options,
      disabled,
      colorScheme,
      direction = "horizontal",
      spacing = direction === "horizontal" ? "24px" : "16px",
      name,
      onChange,
      ...otherProps
    } = props

    const [value, setValue] = useMergeValue(undefined, {
      defaultValue: props.defaultValue,
      value: props.value,
    })

    let radioGroupCss: SerializedStyles
    switch (direction) {
      case "vertical":
        radioGroupCss = applyRadioContainerVertical(spacing)
        break
      case "horizontal":
        radioGroupCss = applyRadioContainerHorizontal(spacing)
        break
    }

    function onChangeValue<T>(v: T, event: ChangeEvent): void {
      const { onChange } = props
      if (v !== value) {
        if (!("value" in props)) {
          setValue(v)
        }
        onChange && onChange(true, event)
      }
    }

    const contextProp = {
      onChangeValue,
      name,
      options,
      disabled,
      value,
      spacing,
      colorScheme,
    }

    return (
      <div css={radioGroupCss} ref={ref} {...otherProps}>
        <RadioGroupContext.Provider value={contextProp}>
          {options && isArray(options)
            ? options.map((option, index) => {
                return typeof option === "string" ? (
                  <Radio key={index} value={option} disabled={disabled}>
                    {option}
                  </Radio>
                ) : (
                  <Radio
                    key={option.value}
                    value={option.value}
                    disabled={disabled || option.disabled}
                  >
                    {option.label}
                  </Radio>
                )
              })
            : children}
        </RadioGroupContext.Provider>
      </div>
    )
  },
)

RadioGroupContext.displayName = "RadioGroupContext"

RadioGroup.displayName = "RadioGroup"
