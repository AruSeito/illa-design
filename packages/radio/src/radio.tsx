/** @jsxImportSource @emotion/react */
import * as React from "react"
import { forwardRef, ChangeEvent, useContext } from "react"
import { RadioProps } from "./interface"
import { RadioGroupContext } from "./radio-group"
import { applyMergeCss, applyRadioSize } from "./style"
import { useMergeValue } from "./hook"

export const Radio = forwardRef<HTMLLabelElement, RadioProps>((props, ref) => {
  const colorScheme = props?.colorScheme ?? "blue"

  const context = useContext(RadioGroupContext);
  const mergeProps = { ...props };
  const { children, disabled, value  } = mergeProps;

  if (context) {
    mergeProps.checked = context?.value === props?.value;
    mergeProps.disabled = !!(context?.disabled || props?.disabled);
  }

  const [currentChecked, setCurrentChecked] = useMergeValue(false, {
    value: mergeProps.checked,
    defaultValue: mergeProps.defaultChecked,
  })

  const onChange = (event: ChangeEvent<HTMLInputElement>): void => {
    const { onChange, value } = mergeProps;
    if (disabled) {
      return;
    }
    if (context) {
      context.onChangeValue && context.onChangeValue(value, event);
    } else if (!('checked' in props) && !currentChecked) {
      setCurrentChecked(true);
    }
    !currentChecked && onChange && onChange(true, event);
  };

  return <label css={applyMergeCss(props)} ref={ref}>
        <input
          type="radio"
          {...(context?.name ? { name: context.name } : {})}
          css={applyRadioSize(colorScheme)}
          value={value||''}
          checked={currentChecked}
          disabled={disabled}
          onChange={(event) => {
            event.persist();
            onChange(event);
          }}
          onClick={(e) => {
            e.stopPropagation();
          }}
        />
        {children}
      </label>

})
