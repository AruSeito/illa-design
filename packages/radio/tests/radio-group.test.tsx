import * as React from "react"
import { Radio, RadioGroup } from "../src"
import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import "@testing-library/jest-dom"

test("RadioGroup renders correctly", () => {
  const { asFragment } = render(
    <RadioGroup data-testid="test-radio-group" aria-label="radio-group" role={"group"}>
      <Radio value="a">a</Radio>
      <Radio value="b">b</Radio>
      <Radio value="c">c</Radio>
    </RadioGroup>,
  )
  // expect(asFragment()).toMatchSnapshot()
  // expect(screen.getAllByRole("group")).toBeInTheDocument()
  // expect(screen.getAllByRole("radio-group")).toBeInTheDocument()
})

test("RadioGroup options renders correctly", () => {
  render(
    <RadioGroup options={["A", "B", "C"]} data-testid="test-radio-group-1">
    </RadioGroup>,
  )
  // expect(screen.getByTestId("test-radio-group-1")).toBeInTheDocument()
})

test("RadioGroup renders with value", () => {
  render(
    <RadioGroup options={["valueA", "valueB", "valueC"]} value={"valueA"}>
    </RadioGroup>,
  )
  expect(screen.getByLabelText("valueA")).toBeChecked()
})

test("RadioGroup renders with click", async () => {
  render(
    <RadioGroup options={["clickA", "clickB", "clickC"]}>
    </RadioGroup>,
  )
  fireEvent.click(screen.getByLabelText("clickB"))
  await waitFor(() => {
    expect(screen.getByLabelText("clickB")).toBeChecked()
  }, {
    timeout: 3000,
  })
})