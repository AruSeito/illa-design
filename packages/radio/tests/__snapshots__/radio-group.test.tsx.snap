// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioGroup renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  vertical-align: middle;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 24px;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  overflow: hidden;
  word-wrap: break-word;
  white-space: nowrap;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  color: #1f1f1f;
  cursor: pointer;
}

.emotion-2 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  border-radius: 50%;
  margin: auto 6px auto auto;
  width: 16px;
  height: 16px;
  border: solid 2px #ebebeb;
  cursor: pointer;
  -webkit-transition: 0.15s all linear;
  transition: 0.15s all linear;
}

.emotion-2:hover {
  border-color: #c2d8ff;
}

.emotion-2:disabled {
  cursor: not-allowed;
  border-color: #ebebeb;
  background-color: #f5f5f5;
}

.emotion-2:checked {
  border: 4px solid #134ae0;
}

.emotion-2:checked:hover {
  border-color: #175ceb;
}

.emotion-2:checked:disabled {
  border-color: #99beff;
}

<div
    class="emotion-0"
  >
    <label
      class="emotion-1"
    >
      <input
        class="emotion-2"
        name="1641367698699"
        type="radio"
        value="a"
      />
      a
    </label>
    <label
      class="emotion-1"
    >
      <input
        class="emotion-2"
        name="1641367698699"
        type="radio"
        value="b"
      />
      b
    </label>
    <label
      class="emotion-1"
    >
      <input
        class="emotion-2"
        name="1641367698699"
        type="radio"
        value="c"
      />
      c
    </label>
  </div>
</DocumentFragment>
`;
